frontend/src/magic/ArrayElement.ts

```diff
@@ -0,0 +1,5 @@
+// evil typescript magic, I stole it from stack overflow
+// don't stare at it too long, you'll go mad
+
+export type ArrayElement<ArrayType extends readonly unknown[]> =
+    ArrayType extends readonly (infer ElementType)[] ? ElementType : never;
```

Comment:
Please briefly, less than 2 paragraphs, explain what this does.

Reply:
I mustn't, I'll go mad surely